lib/snapshot.go

```diff
@@ -717,20 +766,26 @@ func (snap *Snapshot) PrepareAncestralRecord(key string, value []byte, existed b
 	return nil
 }
 
-// FlushAncestralRecords updates the ancestral records after a UtxoView flush.
-// This function should be called in a go-routine after all UtxoView flushes.
-func (snap *Snapshot) FlushAncestralRecords() {
-	glog.V(2).Infof("Snapshot.StartAncestralRecordsFlush: Initiated the flush")
+// FlushAncestralRecordsWithTxn updates the ancestral records after a UtxoView flush.
+// This function should be called in the same badger transaction at the end of UtxoView
+// flushes.
+func (snap *Snapshot) FlushAncestralRecordsWithTxn(txn *badger.Txn) {
+	// If we haven't hit the first snapshot block height yet, don't bother.
+	if !snap.shouldPerformSnapshotOperations() {
+		return
+	}
+	glog.V(2).Infof("Snapshot.FlushAncestralRecords: Initiated the flush")
 
 	// Make sure we've finished all checksum computation before we proceed with the flush.
 	// Since this gets called after all snapshot operations are enqueued after the main db
 	// flush, the order of operations is preserved; however, there could still be some
 	// snapshot worker threads running so we want to wait until they're done.
 	err := snap.Checksum.Wait()
 	if err != nil {
-		glog.Errorf("Snapshot.StartAncestralRecordsFlush: Error while waiting "+
+		glog.Errorf("Snapshot.FlushAncestralRecords: Error while waiting "+
 			"for checksum: (%v)", err)
-		snap.StartAncestralRecordsFlush(false)
+		// TODO: do we need this still?
+		snap.FlushAncestralRecordsWithTxn(txn)
```

Comment:
I'm not sure what this is doing. I'll take a closer look later.

Reply:
It's a recursive call so it's kinda weird.

Reply:
@lazynina I just reviewed this and the instance below it. I think we should just make FlushAncestralRecords return an error in this case rather than doing a recursive call. It seems like the only reason this was done is because they didn't want to change the call signature to return an error maybe.